### PR TITLE
torch.cuda.memory_allocated to return `{}` if not initialized

### LINKED ISSUE #49952: torch.cuda.memory_allocated() doesn't correctly work until the context is initialized
## 🐛 Bug

The manpage https://pytorch.org/docs/stable/cuda.html#torch.cuda.memory_allocated advertises that it's possible to pass an `int` argument but it doesn't work.

And even if I create a device argument it doesn't work correctly in multi-gpu settings - giving me stats for the first device for all other devices.

## To Reproduce

```
python -c 'import torch; print(torch.cuda.memory_allocated(0))'
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/home/stas/anaconda3/envs/main-38/lib/python3.8/site-packages/torch/cuda/memory.py", line 306, in memory_allocated
    return memory_stats(device=device)["allocated_bytes.all.current"]
  File "/home/stas/anaconda3/envs/main-38/lib/python3.8/site-packages/torch/cuda/memory.py", line 187, in memory_stats
    stats = memory_stats_as_nested_dict(device=device)
  File "/home/stas/anaconda3/envs/main-38/lib/python3.8/site-packages/torch/cuda/memory.py", line 197, in memory_stats_as_nested_dict
    return torch._C._cuda_memoryStats(device)
RuntimeError: Invalid device argument.
```

If I make use of the context manager device as a device, then it works:
```
python -c 'import torch; print(torch.cuda.memory_allocated(torch.cuda.device("cuda:0")))'
0
```

It also starts working if I first call `torch.cuda.get_device_name(0))`:

```
python -c 'import torch; print(torch.cuda.get_device_name(0)); print(torch.cuda.memory_allocated(0))'
```

It also seems to just give device 0 stats regardless of what I pass to it, e.g. with 2 gpus:
```
python -c 'import torch; x=torch.ones(10<<20).to(0); y=torch.ones(10).to(1);print([torch.cuda.memory_allocated(torch.cuda.device(f"cuda:{id}")) for id in range(torch.cuda.device_count())])'
[41943040, 41943040]
```
`torch.cuda.memory_allocated` for device 1 returns the same stats as device 0.

If I replace `torch.cuda.device` with `torch.device` it works correctly. But then fails again if I don't do the `to()` calls:

```
python -c 'import torch; x=torch.ones(10<<10); y=torch.ones(10);print([torch.cuda.memory_allocated(torch.device(f"cuda:{id}")) for id in range(torch.cuda.device_count())]);'
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "<string>", line 1, in <listcomp>
  File "/home/stas/anaconda3/envs/main-38/lib/python3.8/site-packages/torch/cuda/memory.py", line 306, in memory_allocated
    return memory_stats(device=device)["allocated_bytes.all.current"]
  File "/home/stas/anaconda3/envs/main-38/lib/python3.8/site-packages/torch/cuda/memory.py", line 187, in memory_stats
    stats = memory_stats_as_nested_dict(device=device)
  File "/home/stas/anaconda3/envs/main-38/lib/python3.8/site-packages/torch/cuda/memory.py", line 197, in memory_stats_as_nested_dict
    return torch._C._cuda_memoryStats(device)
RuntimeError: Invalid device argument.
```

A pattern appears that this function only works once some operation on the desired device has been performed, and then it reports correctly only for that device and not others.

I'm seeing other ops having the same problem: e.g. [this](https://github.com/NVIDIA/apex/issues/1022) and [this](https://github.com/pytorch/pytorch/issues/21819) and the fix is to run some other safe op like  `torch.cuda.set_device` which then fixes these unsafe ops.

It appears that the only way I can take a snapshot of memory allocations per device is with context manager:
```
            per_device_memory = []
            for id in range(torch.cuda.device_count()):
                with torch.cuda.device(id):
                    per_device_memory.append(torch.cuda.memory_allocated())
```

and this is not possible:
```
per_device_memory = [torch.cuda.memory_allocated(id) for id in range(torch.cuda.device_count())]
```
which is very odd.

## Expected behavior

1. Either the documentation is wrong, or there is a bug and it should return the memory usage for an `int` device number. 
2. The behavior seems to be inconsistent depending on what cuda functions are called prior to this function
3. `torch.device` and `torch.cuda.device` seem to be interchangeable in some situations but not others
4. it must not report memory usage for 0th device when it's being asked to report memory usage for other devices

## Environment
```
PyTorch version: 1.8.0.dev20201219+cu110
Is debug build: False
CUDA used to build PyTorch: 11.0
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.1 LTS (x86_64)
GCC version: (Ubuntu 9.3.0-17ubuntu1~20.04) 9.3.0
Clang version: 10.0.0-4ubuntu1 
CMake version: version 3.18.2

Python version: 3.8 (64-bit runtime)
Is CUDA available: True
CUDA runtime version: Could not collect
GPU models and configuration:
GPU 0: GeForce GTX 1070 Ti
GPU 1: GeForce RTX 3090

Nvidia driver version: 455.45.01
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.7.6.5
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.0.4
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.0.4
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.0.4
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.0.4
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.0.4
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.0.4
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.0.4
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] numpy==1.19.4
[pip3] pytorch-lightning==1.1.1rc0
[pip3] pytorch-memlab==0.2.2
[pip3] torch==1.8.0.dev20201219+cu110
[pip3] torchtext==0.6.0
[pip3] torchvision==0.9.0a0+f80b83e
[conda] blas                      1.0                         mkl
[conda] magma-cuda111             2.5.2                         1    pytorch
[conda] mkl                       2020.2                      256
[conda] mkl-include               2020.2                      256
[conda] mkl-service               2.3.0            py38he904b0f_0
[conda] mkl_fft                   1.2.0            py38h23d657b_0
[conda] mkl_random                1.1.1            py38h0573a6f_0
[conda] numpy                     1.19.4                   pypi_0    pypi
[conda] pytorch-lightning         1.1.1rc0                  dev_0    <develop>
[conda] pytorch-memlab            0.2.2                    pypi_0    pypi
[conda] torch                     1.8.0.dev20201219+cu110          pypi_0    pypi
[conda] torchtext                 0.6.0                    pypi_0    pypi
[conda] torchvision               0.9.0a0+f80b83e           dev_0    <develop>
```

cc @ezyang @gchanan @zou3519 @bdhirsh @jbschlosser @ngimel @jlin27 @mruberry

### PR DESCRIPTION
**Pull request by malfet**
Fixes #49952

### CODE REVIEW

**Comment by ngimel** [comment]
Thanks for the fix!
Can we add tests? Also, another case reported in #49952 still does not work (reports equal memory usage for both gpus)
```
python -c 'import torch; x=torch.ones(10<<20).to(0); y=torch.ones(10).to(1);print([torch.cuda.memory_allocated(torch.cuda.device(id)) for id in range(torch.cuda.device_count())])'
```
Weirdly, it works if `torch.cuda.device(id)` is replaced by `id`.

**Comment by malfet** [comment]
> Weirdly, it works if `torch.cuda.device(id)` is replaced by `id`.
Fixed, although according to type annotations, this function is not supposed to be called with `torch.cuda.device`, which despite the name is not inherited from `torch.device`, filed https://github.com/pytorch/pytorch/issues/51224 to track that.

**Comment by ngimel** [comment]
Awesome, can you please add tests?

File: test/test_cuda.py [hunk 1]
```diff
@@ -3496,6 +3496,15 @@ class TestNamedTupleInput_1(NamedTuple):
             self.assertEqual(expected_a, x.a)
             self.assertEqual(expected_b, x.b)
 
+    @unittest.skipIf(not TEST_MULTIGPU, "Test needs multiple GPUs")
+    def test_cuda_device_memory_allocated(self):
+        from torch.cuda import memory_allocated
+        device_count = torch.cuda.device_count()
+        current_alloc = [memory_allocated(idx) for idx in range(device_count)]
+        x = torch.ones(10, device="cuda:0")
+        self.assertTrue(torch.cuda.memory_allocated(0) > current_alloc[0])
```
Comment (ngimel) [inline test/test_cuda.py:3505]: nit: you don't need torch.cuda here.

**Review by ngimel — APPROVED** [review]

**Pull request closed without merging**